File: src/7_2.py
import re

from sys import stdin


def main():
    variables = {}

    def get_value(value):
        if isinstance(value, int):
            return value
        else:
            return get_variable_value(value)

    cache = {}

    @cached(cache)
    def get_variable_value(name):
        return variables[name]()

    for line in stdin:
        target, function = parse(get_value, line)
        variables[target] = function
    a = get_value('a')
    variables['b'] = lambda: a
    cache.clear()
    print(get_value('a'))


def cached(cache):
    def wrap(function):
        def wrapped(*args):
            result = cache.get(args)
            if result is not None:
                return result
            result = function(*args)
            cache[args] = result
            return result
        return wrapped
    return wrap


ASSIGN = re.compile(r'^(?P<value>\w+) -> (?P<target>\w+)$')
UNARY = re.compile(r'^(?P<operation>NOT) (?P<value>\w+) -> (?P<target>\w+)$')
BINARY = re.compile(r'^(?P<lhs>\w+) (?P<operation>AND|OR|LSHIFT|RSHIFT)'
                    r' (?P<rhs>\w+) -> (?P<target>\w+)$')


def parse(get_value, line):
    for r, f in ((ASSIGN, on_assign), (UNARY, on_unary), (BINARY, on_binary)):
        match = r.match(line)
        if match:
            return f(get_value, match)


def on_assign(get_value, match):
    value = get_int_or_str(match.group('value'))
    target = match.group('target')
    return target, lambda: get_value(value)


def on_unary(get_value, match):
    value = get_int_or_str(match.group('value'))
    operation = dict(
        NOT=lambda x: ~x,
    )[match.group('operation')]
    target = match.group('target')
    return target, lambda: operation(get_value(value)) % 65536


def on_binary(get_value, match):
    lhs = get_int_or_str(match.group('lhs'))
    rhs = get_int_or_str(match.group('rhs'))
    operation = dict(
        AND=lambda a, b: a & b,
        OR=lambda a, b: a | b,
        LSHIFT=lambda a, b: a << b,
        RSHIFT=lambda a, b: a >> b,
    )[match.group('operation')]
    target = match.group('target')
    return target, lambda: operation(get_value(lhs), get_value(rhs)) % 65536


def get_int_or_str(value):
    return int(value) if value.isdigit() else value


if __name__ == '__main__':
    main()

File: src/7_2.rs
extern crate regex;

use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use regex::{Regex, Captures};

type Variables = HashMap<String, Box<Fn() -> u16>>;
type VariablesArc = Arc<RwLock<Variables>>;
type Cache = HashMap<String, u16>;
type CacheArc = Arc<RwLock<Cache>>;

struct Parser {
    variables: VariablesArc,
    cache: CacheArc,
    assign: Regex,
    unary: Regex,
    binary: Regex,
}

fn main() {
    use std::io::{stdin, BufRead};
    let input = stdin();
    let variables = Arc::new(RwLock::new(Variables::new()));
    let cache = Arc::new(RwLock::new(Cache::new()));
    let parser = Parser::new(variables.clone(), cache.clone());
    for line in input.lock().lines() {
        let (target, function) = parser.parse(&line.unwrap()[..]);
        variables.write().unwrap().insert(target, function);
    }
    let a = variables.read().unwrap()["a"]();
    variables.write().unwrap().insert("b".to_string(), Box::new(move || a));
    cache.write().unwrap().clear();
    let result = variables.read().unwrap()["a"]();
    println!("{}", result);
}

impl Parser {
    pub fn new(variables: VariablesArc, cache: CacheArc) -> Parser {
        Parser {
            variables: variables,
            cache: cache,
            assign: Regex::new(
                r"^(?P<value>\w+) -> (?P<target>\w+)$"
            ).unwrap(),
            unary: Regex::new(
                r"^(?P<operation>NOT) (?P<value>\w+) -> (?P<target>\w+)$"
            ).unwrap(),
            binary: Regex::new(
                r"^(?P<lhs>\w+) (?P<operation>AND|OR|LSHIFT|RSHIFT) (?P<rhs>\w+) -> (?P<target>\w+)$"
            ).unwrap()
        }
    }

    pub fn parse(&self, line: &str) -> (String, Box<Fn() -> u16>) {
        self.try_assign(line)
    }

    fn try_assign(&self, line: &str) -> (String, Box<Fn() -> u16>) {
        match self.assign.captures(line) {
            Some(captures) => self.on_assign(captures),
            None => self.try_unary(line),
        }
    }

    fn try_unary(&self, line: &str) -> (String, Box<Fn() -> u16>) {
        match self.unary.captures(line) {
            Some(captures) => self.on_unary(captures),
            None => self.try_binary(line),
        }
    }

    fn try_binary(&self, line: &str) -> (String, Box<Fn() -> u16>) {
        match self.binary.captures(line) {
            Some(captures) => self.on_binary(captures),
            None => panic!(),
        }
    }

    fn on_assign(&self, captures: Captures) -> (String, Box<Fn() -> u16>) {
        let value = captures.name("value").unwrap();
        let target = captures.name("target").unwrap();
        (target.to_string(), self.make_get_function(value))
    }

    fn on_unary(&self, captures: Captures) -> (String, Box<Fn() -> u16>) {
        let value = captures.name("value").unwrap();
        let operation = captures.name("operation").unwrap();
        let target = captures.name("target").unwrap();
        let get_value = self.make_get_function(value);
        let function: Box<Fn() -> u16> = match operation {
            "NOT" => Box::new(move || !get_value()),
            _ => panic!(),
        };
        (target.to_string(), function)
    }

    fn on_binary(&self, captures: Captures) -> (String, Box<Fn() -> u16>) {
        let lhs = captures.name("lhs").unwrap();
        let rhs = captures.name("rhs").unwrap();
        let operation = captures.name("operation").unwrap();
        let target = captures.name("target").unwrap();
        let get_lhs = self.make_get_function(lhs);
        let get_rhs = self.make_get_function(rhs);
        let function: Box<Fn() -> u16> = match operation {
            "AND" => Box::new(move || get_lhs() & get_rhs()),
            "OR" => Box::new(move || get_lhs() | get_rhs()),
            "LSHIFT" => Box::new(move || get_lhs() << get_rhs()),
            "RSHIFT" => Box::new(move || get_lhs() >> get_rhs()),
            _ => panic!(),
        };
        (target.to_string(), function)
    }

    fn make_get_function(&self, value: &str) -> Box<Fn() -> u16> {
        let result: Box<Fn() -> u16> = match value.parse::<u16>() {
            Ok(number) => Box::new(move || number),
            Err(_) => {
                let variables = self.variables.clone();
                let cache = self.cache.clone();
                let name = value.to_string();
                Box::new(move || {
                    let cached = match cache.read().unwrap().get(&name) {
                        Some(&value) => Some(value),
                        None => None,
                    };
                    match cached {
                        Some(value) => value,
                        None => {
                            let value = variables.read().unwrap()[&name]();
                            cache.write().unwrap().insert(name.clone(), value);
                            value
                        }
                    }
                })
            },
        };
        result
    }
}
